File: python/src/fancy_expr/_parse.py
"""Recursive-descent parser producing a plain-dict AST.

Separable from evaluation on purpose: a host must be able to ask *"is this
expression valid?"* with no data at all, which is what makes a workflow node
carrying a broken expression rejectable at SAVE time rather than discovered
half-way through a run.
"""

from __future__ import annotations

from typing import Any

from ._tokenize import ExprSyntaxError, Token, tokenize

# Precedence, lowest to highest, exactly as GRAMMAR.md lists it:
#   ternary, ||, &&, equality, comparison, + -, * /, unary, primary


class _State:
    __slots__ = ("expression", "i", "tokens")

    def __init__(self, tokens: list[Token], expression: str) -> None:
        self.tokens = tokens
        self.i = 0
        self.expression = expression

    def peek(self) -> Token:
        return self.tokens[self.i]

    def next(self) -> Token:
        tok = self.tokens[self.i]
        self.i += 1
        return tok

    def fail(self, message: str) -> ExprSyntaxError:
        tok = self.peek()
        return ExprSyntaxError(message, tok.start, self.expression)


def parse(expression: str) -> dict[str, Any]:
    """Parse without evaluating. Raises ``ExprSyntaxError`` on anything malformed."""
    s = _State(tokenize(expression), expression)

    # Row 0804. An empty expression is malformed, not an empty value -- a host
    # asking "is this valid?" about a blank field must hear no.
    if s.peek().kind == "eof":
        raise ExprSyntaxError("Empty expression", 0, expression)

    node = _expression(s)

    if s.peek().kind != "eof":
        raise s.fail("Unexpected trailing input")

    return node


def _is_punct(s: _State, *values: str) -> bool:
    tok = s.peek()
    return tok.kind == "punct" and tok.value in values


def _expect_punct(s: _State, value: str) -> None:
    if not _is_punct(s, value):
        raise s.fail(f"Expected {value!r}")
    s.next()


def _expression(s: _State) -> dict[str, Any]:
    return _ternary(s)


def _ternary(s: _State) -> dict[str, Any]:
    cond = _or(s)
    if not _is_punct(s, "?"):
        return cond
    s.next()
    then = _expression(s)
    _expect_punct(s, ":")
    return {"kind": "ternary", "cond": cond, "then": then, "otherwise": _expression(s)}


def _or(s: _State) -> dict[str, Any]:
    left = _and(s)
    while _is_punct(s, "||"):
        s.next()
        left = {"kind": "logical", "op": "||", "left": left, "right": _and(s)}
    return left


def _and(s: _State) -> dict[str, Any]:
    left = _equality(s)
    while _is_punct(s, "&&"):
        s.next()
        left = {"kind": "logical", "op": "&&", "left": left, "right": _equality(s)}
    return left


def _equality(s: _State) -> dict[str, Any]:
    left = _comparison(s)
    while _is_punct(s, "==", "===", "!=", "!=="):
        # ONE operator. Both spellings collapse here, so no runtime can end up
        # with a subtly different `==` from the other two (rows 0501-0503).
        raw = s.next().value
        op = "===" if raw in ("==", "===") else "!=="
        left = {"kind": "binary", "op": op, "left": left, "right": _comparison(s)}
    return left


def _comparison(s: _State) -> dict[str, Any]:
    left = _additive(s)
    while _is_punct(s, "<", "<=", ">", ">="):
        op = s.next().value
        left = {"kind": "binary", "op": op, "left": left, "right": _additive(s)}
    return left


def _additive(s: _State) -> dict[str, Any]:
    left = _multiplicative(s)
    while _is_punct(s, "+", "-"):
        op = s.next().value
        left = {"kind": "binary", "op": op, "left": left, "right": _multiplicative(s)}
    return left


def _multiplicative(s: _State) -> dict[str, Any]:
    left = _unary(s)
    while _is_punct(s, "*", "/"):
        op = s.next().value
        left = {"kind": "binary", "op": op, "left": left, "right": _unary(s)}
    return left


def _unary(s: _State) -> dict[str, Any]:
    if _is_punct(s, "!", "-"):
        op = s.next().value
        return {"kind": "unary", "op": op, "operand": _unary(s)}
    return _primary(s)


def _primary(s: _State) -> dict[str, Any]:
    tok = s.peek()

    if tok.kind == "number" or tok.kind == "string":
        s.next()
        return {"kind": "literal", "value": tok.value}

    if tok.kind == "ident":
        if tok.value == "true":
            s.next()
            return {"kind": "literal", "value": True}
        if tok.value == "false":
            s.next()
            return {"kind": "literal", "value": False}
        if tok.value == "null":
            s.next()
            return {"kind": "literal", "value": None}
        return _path(s)

    if _is_punct(s, "("):
        s.next()
        node = _expression(s)
        # Row 0803. An unclosed paren is malformed; it is never "the rest is
        # implied".
        _expect_punct(s, ")")
        return node

    if _is_punct(s, "["):
        return _array(s)

    if _is_punct(s, "{"):
        return _object(s)

    # Row 0802 lands here: a dangling operator leaves nothing for the operand.
    raise s.fail("Expected an expression")


def _path(s: _State) -> dict[str, Any]:
    root = s.next().value

    # The root is the FIRST step, not a special case beside them. One lookup
    # path is one thing to keep correct; two eventually stop agreeing.
    steps: list[dict[str, Any]] = [{"type": "prop", "name": root}]

    while True:
        if _is_punct(s, "."):
            s.next()
            tok = s.peek()
            if tok.kind != "ident":
                raise s.fail("Expected a property name after '.'")
            s.next()
            steps.append({"type": "prop", "name": tok.value})
            continue

        if _is_punct(s, "["):
            s.next()
            index = _expression(s)
            _expect_punct(s, "]")
            steps.append({"type": "index", "expr": index})
            continue

        # Row 0805, and the security boundary. A `(` after a path is a CALL, and
        # calls do not parse -- which is what lets a host reject one before it
        # is ever saved, rather than sandboxing it at run time.
        if _is_punct(s, "("):
            raise s.fail("Function calls are not permitted")

        return {"kind": "path", "steps": steps}


def _array(s: _State) -> dict[str, Any]:
    _expect_punct(s, "[")
    items: list[dict[str, Any]] = []

    if not _is_punct(s, "]"):
        items.append(_expression(s))
        while _is_punct(s, ","):
            s.next()
            items.append(_expression(s))

    _expect_punct(s, "]")
    return {"kind": "array", "items": items}


def _object(s: _State) -> dict[str, Any]:
    _expect_punct(s, "{")
    entries: list[tuple[str, dict[str, Any]]] = []

    if not _is_punct(s, "}"):
        entries.append(_entry(s))
        while _is_punct(s, ","):
            s.next()
            entries.append(_entry(s))

    _expect_punct(s, "}")
    return {"kind": "object", "entries": entries}


def _entry(s: _State) -> tuple[str, dict[str, Any]]:
    tok = s.peek()
    if tok.kind not in ("ident", "string"):
        raise s.fail("Expected an object key")
    s.next()
    _expect_punct(s, ":")
    return str(tok.value), _expression(s)

File: src/parse.ts
/**
 * The parser. Tokens in, AST out, or it throws.
 *
 * A plain recursive-descent parser following `GRAMMAR.md` exactly, one function
 * per precedence level, lowest first. Deliberately boring: the grammar is the
 * interesting artifact, and a parser that departs from it to be clever is a
 * parser that disagrees with the other two implementations.
 *
 * ## Separable from evaluation ON PURPOSE
 *
 * `parse()` needs no data. That is what lets a host ask *"is this expression
 * valid?"* while a person is still typing it, and reject a node whose
 * expression can never work at SAVE time rather than discovering it mid-run.
 * The whole defect this package replaces was a malformed expression being
 * indistinguishable from an absent value at run time.
 */
import { ExprSyntaxError, tokenize, type Token } from "./tokenize";

export type Node =
  | { kind: "literal"; value: unknown }
  | { kind: "path"; segments: Array<{ name: string } | { expr: Node }> }
  | { kind: "unary"; op: "!" | "-"; operand: Node }
  | { kind: "binary"; op: string; left: Node; right: Node }
  | { kind: "logical"; op: "&&" | "||"; left: Node; right: Node }
  | { kind: "ternary"; test: Node; then: Node; other: Node }
  | { kind: "array"; items: Node[] }
  | { kind: "object"; entries: Array<{ key: string; value: Node }> };

export function parse(source: string): Node {
  const tokens = tokenize(source);
  const state = { tokens, i: 0, source };

  // An empty expression is malformed rather than null. `{{ }}` says nothing,
  // and silently producing a value for it would be the same class of guess this
  // package exists to remove.
  if (peek(state).type === "eof") {
    throw new ExprSyntaxError("Empty expression", source, 0);
  }

  const node = parseTernary(state);

  // Trailing junk is an error. Without this, `in.a in.b` would parse as `in.a`
  // and silently ignore the rest — the single easiest way for a typo to become
  // a wrong answer instead of a message.
  const rest = peek(state);
  if (rest.type !== "eof") {
    throw new ExprSyntaxError(`Unexpected \`${rest.value}\``, source, rest.at);
  }

  return node;
}

type State = { tokens: Token[]; i: number; source: string };

const peek = (s: State): Token => s.tokens[s.i];
const next = (s: State): Token => s.tokens[s.i++];

function isPunct(s: State, ...values: string[]): boolean {
  const t = peek(s);
  return t.type === "punct" && values.includes(t.value as string);
}

function eat(s: State, value: string): boolean {
  if (isPunct(s, value)) {
    s.i += 1;
    return true;
  }
  return false;
}

function expect(s: State, value: string): void {
  if (!eat(s, value)) {
    const t = peek(s);
    throw new ExprSyntaxError(`Expected \`${value}\``, s.source, t.at);
  }
}

function parseTernary(s: State): Node {
  const test = parseOr(s);
  if (!eat(s, "?")) return test;

  const then = parseTernary(s);
  expect(s, ":");
  const other = parseTernary(s);
  return { kind: "ternary", test, then, other };
}

function parseOr(s: State): Node {
  let left = parseAnd(s);
  while (isPunct(s, "||")) {
    s.i += 1;
    left = { kind: "logical", op: "||", left, right: parseAnd(s) };
  }
  return left;
}

function parseAnd(s: State): Node {
  let left = parseEquality(s);
  while (isPunct(s, "&&")) {
    s.i += 1;
    left = { kind: "logical", op: "&&", left, right: parseEquality(s) };
  }
  return left;
}

function parseEquality(s: State): Node {
  let left = parseComparison(s);
  while (isPunct(s, "===", "==", "!==", "!=")) {
    // `==` and `===` collapse to ONE operator here, and so do `!=` and `!==`.
    // Both spellings are accepted because authors write both; neither coerces.
    // Two equality operators that differ subtly is a language that generates
    // bug reports, and there are three runtimes to keep in step.
    const raw = next(s).value as string;
    const op = raw.startsWith("!") ? "!==" : "===";
    left = { kind: "binary", op, left, right: parseComparison(s) };
  }
  return left;
}

function parseComparison(s: State): Node {
  let left = parseAdditive(s);
  while (isPunct(s, "<", "<=", ">", ">=")) {
    const op = next(s).value as string;
    left = { kind: "binary", op, left, right: parseAdditive(s) };
  }
  return left;
}

function parseAdditive(s: State): Node {
  let left = parseMultiplicative(s);
  while (isPunct(s, "+", "-")) {
    const op = next(s).value as string;
    left = { kind: "binary", op, left, right: parseMultiplicative(s) };
  }
  return left;
}

function parseMultiplicative(s: State): Node {
  let left = parseUnary(s);
  while (isPunct(s, "*", "/")) {
    const op = next(s).value as string;
    left = { kind: "binary", op, left, right: parseUnary(s) };
  }
  return left;
}

function parseUnary(s: State): Node {
  if (isPunct(s, "!", "-")) {
    const op = next(s).value as "!" | "-";
    return { kind: "unary", op, operand: parseUnary(s) };
  }
  return parsePrimary(s);
}

function parsePrimary(s: State): Node {
  const t = peek(s);

  if (t.type === "number" || t.type === "string") {
    s.i += 1;
    return { kind: "literal", value: t.value };
  }

  if (t.type === "ident") {
    const name = t.value as string;
    if (name === "true" || name === "false" || name === "null") {
      s.i += 1;
      return { kind: "literal", value: name === "null" ? null : name === "true" };
    }
    return parsePath(s);
  }

  if (eat(s, "(")) {
    const inner = parseTernary(s);
    expect(s, ")");
    return inner;
  }

  if (eat(s, "[")) {
    const items: Node[] = [];
    if (!isPunct(s, "]")) {
      do {
        items.push(parseTernary(s));
      } while (eat(s, ","));
    }
    expect(s, "]");
    return { kind: "array", items };
  }

  if (eat(s, "{")) {
    const entries: Array<{ key: string; value: Node }> = [];
    if (!isPunct(s, "}")) {
      do {
        const k = next(s);
        if (k.type !== "ident" && k.type !== "string") {
          throw new ExprSyntaxError("Expected an object key", s.source, k.at);
        }
        expect(s, ":");
        entries.push({ key: String(k.value), value: parseTernary(s) });
      } while (eat(s, ","));
    }
    expect(s, "}");
    return { kind: "object", entries };
  }

  throw new ExprSyntaxError(`Unexpected \`${t.value || "end of expression"}\``, s.source, t.at);
}

function parsePath(s: State): Node {
  const segments: Array<{ name: string } | { expr: Node }> = [];
  const head = next(s);
  segments.push({ name: head.value as string });

  for (;;) {
    if (eat(s, ".")) {
      const seg = next(s);
      if (seg.type !== "ident") {
        throw new ExprSyntaxError("Expected a path segment", s.source, seg.at);
      }
      segments.push({ name: seg.value as string });
      continue;
    }

    if (eat(s, "[")) {
      const idx = parseTernary(s);
      expect(s, "]");
      segments.push({ expr: idx });
      continue;
    }

    // A `(` here would be a function call. Refused DELIBERATELY: this grammar
    // has no calls, because these expressions arrive from end users and from
    // agents, over the wire. Sandboxing is a security property, not a style.
    if (isPunct(s, "(")) {
      throw new ExprSyntaxError(
        "Function calls are not supported — this is an expression grammar, not a language",
        s.source,
        peek(s).at,
      );
    }

    return { kind: "path", segments };
  }
}
